FaunaDB.Client/Types/Decoder.cs

```diff
@@ -0,0 +1,493 @@
+
+        static object DecodeIntern(Value value, Type dstType)
+        {
+            if (typeof(object) != dstType && dstType.IsAssignableFrom(value.GetType()))
+                return value;
+
+            switch (Type.GetTypeCode(dstType))
+            {
+                case TypeCode.String:
+                    return ToString(value);
+
+                case TypeCode.Boolean:
+                    return ToBoolean(value);
+
+                case TypeCode.DateTime:
+                    return ToDateTime(value);
+
+                case TypeCode.SByte:
+                    return Convert.ToSByte(ToLong<sbyte>(value));
+                case TypeCode.Int16:
+                    return Convert.ToInt16(ToLong<short>(value));
+                case TypeCode.Int32:
+                    return Convert.ToInt32(ToLong<int>(value));
+                case TypeCode.Int64:
+                    return ToLong<long>(value);
+
+                case TypeCode.Char:
+                    return Convert.ToChar(ToLong<char>(value));
+                case TypeCode.Byte:
+                    return Convert.ToByte(ToLong<byte>(value));
+                case TypeCode.UInt16:
+                    return Convert.ToUInt16(ToLong<ushort>(value));
+                case TypeCode.UInt32:
+                    return Convert.ToUInt32(ToLong<uint>(value));
+                case TypeCode.UInt64:
+                    return Convert.ToUInt64(ToLong<ulong>(value));
+
+                case TypeCode.Single:
+                    return Convert.ToSingle(ToDouble<float>(value));
+                case TypeCode.Double:
+                    return ToDouble<double>(value);
+                case TypeCode.Decimal:
+                    return Convert.ToDecimal(ToDouble<decimal>(value));
+
+                case TypeCode.DBNull:
+                case TypeCode.Empty:
+                    return null;
+
+                case TypeCode.Object:
+                    if (typeof(byte[]) == dstType && value is BytesV)
+                        return ToByteArray(value);
+
+                    if (dstType.IsArray && dstType.HasElementType)
+                        return FromArray(value, dstType);
+
+                    if (typeof(IDictionary).IsAssignableFrom(dstType) || (dstType.IsGenericType && dstType.Is(typeof(IDictionary<,>))))
+                        return FromDictionary(value, dstType);
+
+                    if (typeof(IList).IsAssignableFrom(dstType) || (dstType.IsGenericType && dstType.Is(typeof(IEnumerable<>))))
+                        return FromEnumerable(value, dstType);
+
+                    return FromObject(value, dstType);
+            }
+
+            throw new InvalidOperationException($"Cannot convert {value} to {dstType}");
+        }
+
+        static IList FromEnumerable(Value value, Type type)
+        {
+            if (!(value is ArrayV))
+                throw new InvalidOperationException($"Cannot convert `{value}` to {type}");
+
+            if (!type.IsGenericType)
```

Comment:
See the note below in `FromDictionary` - should we assert that the type is assignable from a base interface type?